### PR TITLE
licencjat v3

### PR DESCRIPTION
**Pull request by nawiju**

### CODE REVIEW

File: docs/licencjat.tex [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+    urlcolor=blue
+}
+
+% Dane magistranta:
+\autori{Szymon Kozłowski}{448304}
+\autorii{Gustaw Blachowski}{448194}
+\autoriii{Kamil Dybek}{448224}
+\autoriv{Natalia Junkiert}{448267}
+
+\title{Using machine learning models for processing the data presented to user by mobile devices.}
```
Comment (GustawB) [inline docs/licencjat.tex:49]:
```suggestion
\title{Using machine learning models for processing the data presented to the user by mobile devices.}
```

File: docs/licencjat.tex [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        \hline
+    \end{tabular}
+    \caption{Example of dataset format representing screen content.}
+    \label{tab:dataset_example}
+\end{table}
+
+Additional requirement is that the screen content processing will be performed exclusively on the end device to mitigate potential privacy concerns.
+
+\section{Project goals}
+\begin{enumerate}
```
Comment (GustawB) [inline docs/licencjat.tex:222]: Our promotor requested some introduction here.

**Review by GustawB — CHANGES_REQUESTED** [review]
There are many empty pages. Are they necessary/generated by the compiler?

**Comment by nawiju** [comment]
> There are many empty pages. Are they necessary/generated by the compiler?

It was an issue with the latex format itself; fixed it

File: docs/licencjat.tex [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+
+\chapter{Machine learning and the dangers associated with it}
+% \subsection{Benchmark}
+% Benchmarking is the process of running a set of, among others, computer programs against a set of tests to assess their relative performance or precision % \cite{benchmark}.
+
+Over the past several years, artificial intelligence (AI) has been widely sensationalized in the media. Amid the promises of a utopian future, with self-driving cars and intelligent virtual assistants that dominate the headlines, concerns about AI are also growing. Many fear a future in which human labour has been made obsolete by automation and AI \cite{francuz}. Privacy concerns are also mounting, as AI models are often trained on vast datasets that may include sensitive information such as healthcare records, biometric data for facial recognition, and financial details — sometimes collected without consent \cite{ibm_privacy}.
```
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:245]: changed it to "discussed"

File: docs/licencjat.tex [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+
+\subsection{Deep Learning}
+Deep learning is a subset of machine learning wherein multilayered neural networks, called deep neural networks, are utilized to learn increasingly meaningful representations with each successive layer as seen in ~\ref{fig:nn_simple}; each representation is increasingly different from the original and more useful to determining the result. Traditionally, machine learning models focus on learning one or two layers of representations of the data \cite{francuz}, deep learning employs at least three layers, and typically hundreds or thousands of layers to train the models \cite{ibm_dl}.
```
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:270]: done

File: docs/licencjat.tex [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+
+% \subsubsection{Relevant NLP models}
+
+\section{Should We Be Afraid of AI? - Assessing the Risks and Ethical Implications of Artificial Intelligence}
+
+% nie wydaje mi się że pisanie o adversarial attacks czy accuracy concerns w tym miejscu jest potrzebne. adversarial attacks bo nie robimy czegos co moze byc tak wykorzystane. accuracy bo bedziemy o tym pisać w kolejnych sekcjach i nie ma co od razu kota z worka wyciągać
+
+As artificial intelligence becomes increasingly complex and integrated into our daily lives, the voices raising concerns about its dangers grow louder. Some express concern about the risks of excessive surveillance and privacy erosion, envisioning a future where AI systems are deeply intertwined with our surroundings, essentially hearing what we hear and seeing what we see \cite{sroka_vid}. Others highlight the environmental challenges tied to large-scale AI deployment, such as rising energy and water consumption, along with the need for rare materials in microchip production \cite{ai_env_concerns}. This section will focus on assessing the risks of artificial intelligence and how our solution will take them into account.
```
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:304]: fixed it

File: docs/licencjat.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:245]: This book was written by [François Chollet](https://en.wikipedia.org/wiki/Fran%C3%A7ois_Chollet), who i think is a reliable source. [Here](https://sourestdeeds.github.io/pdf/Deep%20Learning%20with%20Python.pdf) is the link

File: docs/licencjat.tex [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+
+\section{Understanding the difference artificial intelligence, machine learning and deep learning}
+Artificial intelligence (AI), machine learning (ML) and deep learning (DL) are terms often mistakenly used interchangeably to refer to the development of systems capable of performing tasks typically requiring human intelligence such as decision making and speech recognition \cite{ibm_ai}. AI is the umbrella term encompassing among others, machine learning and deep learning, as well as other approaches \cite{francuz}. Machine learning is a subset of AI, in which systems are able to learn and adapt without explicit rules \cite{ibm_ai}. Deep learning is a type of machine learning utilizing neural networks. This hierarchy is depicted in the image below.
```
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:248]: i will resolve this comment as it repeats a different comment

File: docs/licencjat.tex [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+
+\subsection{Machine Learning}
+On the other hand, machine learning is a branch of artificial intelligence in which the machine is trained rather than explicitly programmed, by making inferences from the input data it is presented with \cite{francuz}. Two major types of machine learning are supervised and unsupervised learning. 
+
+Supervised learning refers to an approach that relies on labeled datasets to train or "supervise" the model. The model is provided with input data along with the correct output, allowing it to learn by example. The model analyzes the relationship between the input features and the corresponding labels, gradually improving its ability to make accurate predictions. Supervised learning is commonly used for tasks such as classification, where data is sorted into categories, and regression, where numerical values are predicted based on patterns in the data. This method is widely applied in real-world scenarios like email spam detection, image recognition, and sales forecasting.
```
Comment (SzymonKozl) [inline docs/licencjat.tex:263]: "corresponding labels" -> please change to more general formulation as, as you mentioned below, under supervised learning term we have tasks other that classification too

File: docs/licencjat.tex [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+    \label{fig:nn_simple}
+\end{figure}
+
+The transformation implemented by a layer is defined (parametrized) by its \textit{weights}, which are numerical parameters. Learning involves adjusting these weights to ensure the network accurately maps inputs to their corresponding targets. A deep neural network can have millions of parameters, making it challenging to find the right values since changing one parameter affects the others. To guide this process, a \textit{loss function} measures how far the network's predictions deviate from the expected results, providing a score that reflects its performance. Deep learning relies on using the loss score as feedback to adjust the network's weights, guided by the optimizer using the backpropagation algorithm. Initially, the weights are random, resulting in poor predictions and a high loss score. With each example, the optimizer tweaks the weights to reduce the loss. Repeating this process across many examples gradually minimizes the loss, producing a trained network that closely matches its target outputs. This process is exemplified in ~\ref{fig:nn_function} \cite{francuz}.
```
[the remark this answers is not in the document]
Comment (SzymonKozl) [reply docs/licencjat.tex:279]: do you think that deep dive on param initialization techniques is required here?

**Review by SzymonKozl — CHANGES_REQUESTED** [review]
Gj,
In overall, I miss the explanation of LLM-related terms, as those are the core of our project. I think we should describe terms:
- LLM (and/or transformer)
- fine-tuning
- overfitting (?)
- NER
- quantization
- possibly LoRA (if we want to describe llama fine-tuning in details)
- curriculum learining (?)
Maybe some of these terms should be described in other sections, idk.

Additionally, why github diff shows me that the whle licencjat.tex was added in this PR?

File: docs/licencjat.tex (hunk 8 above)
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:279]: maybe we should include it in a later chapter, when we describe the approaches we considered?

File: docs/licencjat.tex (hunk 2 above)
Comment (nawiju) [reply docs/licencjat.tex:223]: right, added it

**Comment by nawiju** [comment]
> Gj, In overall, I miss the explanation of LLM-related terms, as those are the core of our project. I think we should describe terms:
> 
> * LLM (and/or transformer)
> * fine-tuning
> * overfitting (?)
> * NER
> * quantization
> * possibly LoRA (if we want to describe llama fine-tuning in details)
> * curriculum learining (?)
>   Maybe some of these terms should be described in other sections, idk.

ok, fair point, I will add some of those, but I think curriculum learning, lora and quantization should be explained in later sections. if we describe it here, I think it will feel like reading an encyclopedia while we should introduce these terms when they become relevant

> 
> Additionally, why github diff shows me that the whle licencjat.tex was added in this PR?
idk

File: docs/licencjat.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:245]: fixed citations, added page numbers, I will format the citations later on

File: docs/licencjat.tex [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+
+\begin{figure}
+    \centering
+    \includegraphics[width=0.5\linewidth]{bachelor_images/nn_function.png}
+    \caption{A visualization of how a Deep Learning model works \cite{francuz}}
```
[the remark this answers is not in the document]
Comment (nawiju) [reply docs/licencjat.tex:284]: I'll leave this diagram since I think it does provide a good overview of how dl works but I'll add a separate section on transformers and I'll add other diagrams

File: docs/licencjat.tex [hunk 10]
```diff
@@ -0,0 +1,808 @@
...
+
+% \subsection{Assessing the model - benchmarking}
+% Benchmarking is the process of running a set of, among others, computer programs against a set of tests to assess their relative performance or precision \cite{benchmark}. Benchmarks are employed to evaluate and quantify the efficiency, accuracy, speed, and other metrics of a machine learning model or pipeline, thereby enabling the selection of the best overall solution to our problem.
+
+
+\subsection{Transformers}
+Transformers are deep learning models introduced in the 2017 paper "Attention Is All You Need" by Vaswani et al., which have significantly impacted natural language processing and other sequential data tasks. Unlike traditional recurrent neural networks (RNNs), transformers utilize self-attention mechanisms to process input data in parallel, enhancing efficiency and scalability. This architecture has become foundational in models like BERT, and GPT \cite{medium_t}.
```
Comment (SzymonKozl) [inline docs/licencjat.tex:295]: I think a reference to this paper would be nice here

File: docs/licencjat.tex [hunk 11]
```diff
@@ -0,0 +1,808 @@
...
+\subsubsection{Model architecture}
+The proposed architecture has a encoder-decoder structure wherein “the encoder maps an input sequence of symbol representations $ (x_1, … , x_n) $ to a sequence of continuous representations $ z = (z_1, … , z_n) $”. The decoder takes $ z $ and outputs a sequence $ (y_1, … , y_n) $. At every step, the model utilizes previously-generated symbols as additional input when creating the next output. This is shown in figure ~\ref{fig:transformers_fig}.
```
Comment (SzymonKozl) [inline docs/licencjat.tex:297]: Why quotation marks here? If this is the direct citation, please rephrase it

File: docs/licencjat.tex [hunk 12]
```diff
@@ -0,0 +1,808 @@
...
+\subsubsection{Model architecture}
+The proposed architecture has a encoder-decoder structure wherein “the encoder maps an input sequence of symbol representations $ (x_1, … , x_n) $ to a sequence of continuous representations $ z = (z_1, … , z_n) $”. The decoder takes $ z $ and outputs a sequence $ (y_1, … , y_n) $. At every step, the model utilizes previously-generated symbols as additional input when creating the next output. This is shown in figure ~\ref{fig:transformers_fig}. 
+
+The Encoder consists of 6 identical layers, each with two main parts. The first part is a multi-head self-attention mechanism, and the second is a simple fully connected feed-forward network. Each part uses a residual connection and layer normalization, which helps improve the model's performance. The output of each layer has a fixed size (512 dimensions).
```
Comment (SzymonKozl) [inline docs/licencjat.tex:299]: I think the description of the generalized architecture would be better here than description of exact proposal from Srattention is all you need

File: docs/licencjat.tex [hunk 13]
```diff
@@ -0,0 +1,808 @@
...
+    \centering
+    \includegraphics[width=0.7\linewidth]{bachelor_images/attention_fig.png}
+    \caption{Attention and multi-head attention \cite{attention}}
+    \label{fig:attention_fig}
+\end{figure}
+
+
+\subsection{Quantization}
+Quantization is a technique employed in many fields including machine learning, to reduce the precision of numerical representations within models, typically converting high-precision formats like 32-bit floating point (FP32) to lower-precision formats such as 8-bit integers (INT8). Using integer operations instead of floating-point reduces the computational and memory requirements during inference, thereby making it more efficient and faster, which is an essential benefit for real-time applications. Furthermore, quantization enables deployment on resource-constrained hardware such as smartphones and tablets, while also reducing power consumption due to lighter computational loads. Though this may slightly reduce model accuracy, the trade-off often proves worthwhile in practical applications \cite{ibm_quantization}.
+The two most common quantization techniques are float32 -> float16 and float32 -> int8.
```
Comment (SzymonKozl) [inline docs/licencjat.tex:329]: source?

File: docs/licencjat.tex [hunk 14]
```diff
@@ -0,0 +1,808 @@
...
+\subsubsection{float32 -> float16}
```
Comment (SzymonKozl) [inline docs/licencjat.tex:330]: please shorten or remove this subsection, this chapter is already bloated

File: docs/licencjat.tex [hunk 15]
```diff
@@ -0,0 +1,602 @@
...
+
+\section{Understanding the difference artificial intelligence, machine learning and deep learning}
+Artificial intelligence (AI), machine learning (ML) and deep learning (DL) are terms often mistakenly used interchangeably to refer to the development of systems capable of performing tasks typically requiring human intelligence such as decision making and speech recognition \cite{ibm_ai}. AI is the umbrella term encompassing among others, machine learning and deep learning, as well as other approaches \cite{francuz}. Machine learning is a subset of AI, in which systems are able to learn and adapt without explicit rules \cite{ibm_ai}. Deep learning is a type of machine learning utilizing neural networks. This hierarchy is depicted in the image below.
+
+\begin{figure}
+    \centering
+    \includegraphics[width=0.5\linewidth]{bachelor_images/nvidia_ai_hierarchy.png}
+    \caption{The hierarchy of artificial intelligence, machine learning and deep learning \cite{francuz}}
```
[the remark this answers is not in the document]
Comment (GustawB) [reply docs/licencjat.tex:253]: fair

File: docs/licencjat.tex (hunk 9 above)
[the remark this answers is not in the document]
Comment (GustawB) [reply docs/licencjat.tex:284]: k

File: docs/licencjat.tex [hunk 16]
```diff
@@ -0,0 +1,808 @@
...
+Please refer to the aforementioned paper for more details.
+
+\begin{figure}
+    \centering
+    \includegraphics[width=0.5\linewidth]{bachelor_images/transformer_arch.png}
+    \caption{Transformer Architecture \cite{attention}}
+    \label{fig:transformers_fig}
+\end{figure}
+
+\subsubsection{Attention and multi-head attention}
```
Comment (GustawB) [inline docs/licencjat.tex:312]: I'm not sure if we need this really. While we utilize transformers, I don't think that explaining every detail of this architecture is necessary/

File: docs/licencjat.tex [hunk 17]
```diff
@@ -0,0 +1,808 @@
...
+\subsection{Environmental Concerns}
+AI has a notable carbon footprint due to its increasing energy consumption, particularly during model training and usage \cite{forbes_dl_env}. One study predicts that by 2027, AI-related energy consumption could reach 85–134 TWh, representing nearly 0.5\% of today's global electricity usage. This estimate is based on the energy consumption of the Nvidia A100 servers - the hardware estimated to be used by 95 percent of the A.I. market, and their projected sales in the upcoming years \cite{nyt_el}. Training LLMs typically requires significantly more energy in comparison to making a single prediction on the trained model \cite{sci_dir_comp}, for instance BERT required "the energy of a round-trip transcontinental flight" to train, while GPT-3 emitted 552 metric tons of carbon dioxide which is "the equivalent of 123 gasoline-powered passenger vehicles driven for one year" \cite{sci_am_co2}.
```
Comment (GustawB) [inline docs/licencjat.tex:374]: What study?

**Review by GustawB — CHANGES_REQUESTED** [review]

File: docs/licencjat.tex (hunk 16 above)
Comment (nawiju) [reply docs/licencjat.tex:312]: This is what we agreed on with Szymon (to my understanding). Would you like me to make this section less detailed?

File: docs/licencjat.tex (hunk 11 above)
Comment (nawiju) [reply docs/licencjat.tex:297]: rephrased it

File: docs/licencjat.tex (hunk 10 above)
Comment (nawiju) [reply docs/licencjat.tex:295]: done

File: docs/licencjat.tex (hunk 12 above)
Comment (nawiju) [reply docs/licencjat.tex:299]: i made it more general, what do you think?

Comment (SzymonKozl) [reply docs/licencjat.tex]: im good with it now, closing

**Review by SzymonKozl — APPROVED** [review]
